WIP - Python3 migration

kjerstadius opened the pull request:
This pull request contains what I think is necessary in order to make this plugin Python 3 compatible, while still retaining compatibility with Python 2 so installs on older Kodi versions don't break. The changes are largely based on suggestions by the 2to3 script, but the are a couple of things I've ignored as I don't think they're necessary. There are also some changes that are of the recommended sort and not strictly necessary for Python 3 compatibility (replacement of printf style formatting come to mind).

As I currently don't have a Widevine capable device that is also running Kodi 19 my changes are not tested in a Python 3 environment, but rather a Python 2 one (where it works fine as far as I can tell). As such I wouldn't merge this PR until at least some basic verification has been carried out on Kodi 19. For that reason I'm also marking it a work in progress.

kjerstadius committed:
Replace string modulo formatting with format() method

kjerstadius committed:
Remove unused imports and variable

kjerstadius committed:
Make relative import Python 3 compatible

This way of using relative imports is compatible with Python 2.5+ and 3.
See https://portingguide.readthedocs.io/en/latest/imports.html.

kjerstadius committed:
Handle use of unicode() for Python 3

Solution taken from
https://portingguide.readthedocs.io/en/latest/strings.html#conversion-to-text.

kjerstadius committed:
Replace printf style (modulo) string formatting

Its use is not recommended in Python 3, see
https://docs.python.org/3/library/stdtypes.html?highlight=sprintf#printf-style-string-formatting

kjerstadius committed:
Remove unused and non-existing import

rikardfries commented:
If this need testing I would be happy to help. Have Cmore through ComHem.

kjerstadius commented:
@rikardfries Feel free to give it a go and report back. I'm still on Kodi 18, i.e. Python 2, so there's a reasonable chance that I missed something.

rikardfries commented:
> @rikardfries Feel free to give it a go and report back. I'm still on Kodi 18, i.e. Python 2, so there's a reasonable chance that I missed something.

Can you give me the zip for the addon with python 3 code. 
I have to raise the levels of the addon and the python version,
otherwise the Matrix alpha1 will disable it... 
Then I can try it. :)

kjerstadius commented:
You can get the code from [my fork](https://github.com/kjerstadius/kodi-cmore/tree/python3). The branch is called python3.

rikardfries commented:
> You can get the code from [my fork](https://github.com/kjerstadius/kodi-cmore/tree/python3). The branch is called python3.

When I try to put in the Comhem number I get :

2020-09-29 21:30:15.180 T:7456    ERROR <general>: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                                    - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                                   Error Type: <class 'AttributeError'>
                                                   Error Contents: 'str' object has no attribute 'decode'
                                                   Traceback (most recent call last):
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\default.py", line 5, in <module>
                                                       addon.run()
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\resources\lib\addon.py", line 17, in run
                                                       plugin.run()
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\script.module.routing\lib\routing.py", line 130, in run
                                                       self._dispatch(self.path)
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\script.module.routing\lib\routing.py", line 141, in _dispatch
                                                       view_func(**kwargs)
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\resources\lib\addon.py", line 340, in play
                                                       helper.play(plugin.args['video_id'][0])
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\resources\lib\kodihelper.py", line 211, in play
                                                       login_token = self.get_token()
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\resources\lib\kodihelper.py", line 116, in get_token
                                                       self.set_login_credentials()
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\resources\lib\kodihelper.py", line 106, in set_login_credentials
                                                       return self.set_tv_provider_credentials()
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\resources\lib\kodihelper.py", line 137, in set_tv_provider_credentials
                                                       username = self.get_user_input(username_type)
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\resources\lib\kodihelper.py", line 72, in get_user_input
                                                       query = keyboard.getText().decode('utf-8')
                                                   AttributeError: 'str' object has no attribute 'decode'
                                                   -->End of Python script error report<--
                                                   
2020-09-29 21:30:15.510 T:7456     INFO <general>: Python interpreter stopped
2020-09-29 21:30:15.527 T:4544    ERROR <general>: Playlist Player: skipping unplayable item: 0, path [plugin://plugin.video.cmore/play?video_id=3355874]
2020-09-29 21:30:27.315 T:12972 WARNING <general>: xbmc.translatePath is deprecated and might be removed in future kodi versions. Please use xbmcvfs.translatePath instead.
2020-09-29 21:30:42.378 T:12972 WARNING <general>: xbmc.translatePath is deprecated and might be removed in future kodi versions. Please use xbmcvfs.translatePath instead.
2020-09-29 21:30:57.443 T:12972 WARNING <general>: xbmc.translatePath is deprecated and might be removed in future kodi versions. Please use xbmcvfs.translatePath instead.
2020-09-29 21:31:02.311 T:11668    INFO <general>: initializing python engine.
2020-09-29 21:31:02.904 T:11668 WARNING <general>: xbmc.translatePath is deprecated and might be removed in future kodi versions. Please use xbmcvfs.translatePath instead.
2020-09-29 21:31:02.904 T:11668 WARNING <general>: xbmc.translatePath is deprecated and might be removed in future kodi versions. Please use xbmcvfs.translatePath instead.
2020-09-29 21:31:10.375 T:11668   ERROR <general>: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                                    - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                                   Error Type: <class 'requests.exceptions.ConnectionError'>
                                                   Error Contents: HTTPSConnectionPool(host='account-june.b17g.services', port=443): Max retries exceeded with url: /api/?client=cmore-kodi (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x000002F5C8E30A90>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))
                                                   Traceback (most recent call last):
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\script.module.urllib3\lib\urllib3\connection.py", line 156, in _new_conn
                                                       conn = connection.create_connection(
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\script.module.urllib3\lib\urllib3\util\connection.py", line 61, in create_connection
                                                       for res in socket.getaddrinfo(host, port, family, socket.SOCK_STREAM):
                                                     File "C:\Program Files\Kodi\system\python\Lib\socket.py", line 918, in getaddrinfo
                                                       for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
                                                   socket.gaierror: [Errno 11001] getaddrinfo failed
                                                   
                                                   During handling of the above exception, another exception occurred:
                                                   
                                                   Traceback (most recent call last):
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\script.module.urllib3\lib\urllib3\connectionpool.py", line 665, in urlopen
                                                       httplib_response = self._make_request(
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\script.module.urllib3\lib\urllib3\connectionpool.py", line 376, in _make_request
                                                       self._validate_conn(conn)
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\script.module.urllib3\lib\urllib3\connectionpool.py", line 994, in _validate_conn
                                                       conn.connect()
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\script.module.urllib3\lib\urllib3\connection.py", line 300, in connect
                                                       conn = self._new_conn()
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\script.module.urllib3\lib\urllib3\connection.py", line 168, in _new_conn
                                                       raise NewConnectionError(
                                                   urllib3.exceptions.NewConnectionError: <urllib3.connection.VerifiedHTTPSConnection object at 0x000002F5C8E30A90>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed
                                                   
                                                   During handling of the above exception, another exception occurred:
                                                   
                                                   Traceback (most recent call last):
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\script.module.requests\lib\requests\adapters.py", line 439, in send
                                                       resp = conn.urlopen(
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\script.module.urllib3\lib\urllib3\connectionpool.py", line 719, in urlopen
                                                       retries = retries.increment(
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\script.module.urllib3\lib\urllib3\util\retry.py", line 436, in increment
                                                       raise MaxRetryError(_pool, url, error or ResponseError(cause))
                                                   urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='account-june.b17g.services', port=443): Max retries exceeded with url: /api/?client=cmore-kodi (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x000002F5C8E30A90>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))
                                                   
                                                   During handling of the above exception, another exception occurred:
                                                   
                                                   Traceback (most recent call last):
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\default.py", line 5, in <module>
                                                       addon.run()
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\resources\lib\addon.py", line 17, in run
                                                       plugin.run()
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\script.module.routing\lib\routing.py", line 130, in run
                                                       self._dispatch(self.path)
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\script.module.routing\lib\routing.py", line 141, in _dispatch
                                                       view_func(**kwargs)
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\resources\lib\addon.py", line 340, in play
                                                       helper.play(plugin.args['video_id'][0])
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\resources\lib\kodihelper.py", line 211, in play
                                                       login_token = self.get_token()
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\resources\lib\kodihelper.py", line 120, in get_token
                                                       login_data = self.c.login(username, password, operator)
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\resources\lib\cmore.py", line 161, in login
                                                       credentials = self.make_request(url, method, params=params, payload=json.dumps(payload), headers=headers)
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\resources\lib\cmore.py", line 77, in make_request
                                                       req = self.http_session.post(url, params=params, data=payload, headers=headers)
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\script.module.requests\lib\requests\sessions.py", line 581, in post
                                                       return self.request('POST', url, data=data, json=json, **kwargs)
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\script.module.requests\lib\requests\sessions.py", line 533, in request
                                                       resp = self.send(prep, **send_kwargs)
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\script.module.requests\lib\requests\sessions.py", line 646, in send
                                                       r = adapter.send(request, **kwargs)
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\script.module.requests\lib\requests\adapters.py", line 516, in send
                                                       raise ConnectionError(e, request=request)
                                                   requests.exceptions.ConnectionError: HTTPSConnectionPool(host='account-june.b17g.services', port=443): Max retries exceeded with url: /api/?client=cmore-kodi (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x000002F5C8E30A90>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))
                                                   -->End of Python script error report<--
                                                   
2020-09-29 21:31:10.698 T:11668    INFO <general>: Python interpreter stopped

kjerstadius committed:
Remove decode() call

kjerstadius commented:
Hi, sorry for the slow response, but I just had a look at this and I think I identified the issue. I just pushed an update, so if you feel like it, please give another shot.

rikardfries commented:
Thanks for helping me.
i change addon.xml to have python 3.0.0.
Then install the addon.
When I try to put in the Comhem number I get a new error :

2020-10-17 16:25:15.264 T:9976     INFO <general>: CPythonInvoker(26, C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\default.py): script successfully run
2020-10-17 16:25:15.501 T:9976     INFO <general>: Python interpreter stopped
2020-10-17 16:25:24.072 T:7932  WARNING <general>: xbmc.translatePath is deprecated and might be removed in future kodi versions. Please use xbmcvfs.translatePath instead.
2020-10-17 16:25:26.077 T:1552     INFO <general>: Loading skin file: DialogAddonSettings.xml, load type: KEEP_IN_MEMORY
2020-10-17 16:25:39.086 T:7932  WARNING <general>: xbmc.translatePath is deprecated and might be removed in future kodi versions. Please use xbmcvfs.translatePath instead.
2020-10-17 16:25:54.109 T:7932  WARNING <general>: xbmc.translatePath is deprecated and might be removed in future kodi versions. Please use xbmcvfs.translatePath instead.
2020-10-17 16:25:54.283 T:2060     INFO <general>: initializing python engine.
2020-10-17 16:25:54.474 T:2060  WARNING <general>: xbmc.translatePath is deprecated and might be removed in future kodi versions. Please use xbmcvfs.translatePath instead.
2020-10-17 16:25:54.474 T:2060  WARNING <general>: xbmc.translatePath is deprecated and might be removed in future kodi versions. Please use xbmcvfs.translatePath instead.
2020-10-17 16:25:54.783 T:2060     INFO <general>: CPythonInvoker(27, C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\default.py): script successfully run
2020-10-17 16:25:55.039 T:2060     INFO <general>: Python interpreter stopped
2020-10-17 16:26:01.542 T:14224    INFO <general>: initializing python engine.
2020-10-17 16:26:01.755 T:14224 WARNING <general>: xbmc.translatePath is deprecated and might be removed in future kodi versions. Please use xbmcvfs.translatePath instead.
2020-10-17 16:26:01.755 T:14224 WARNING <general>: xbmc.translatePath is deprecated and might be removed in future kodi versions. Please use xbmcvfs.translatePath instead.
2020-10-17 16:26:02.048 T:14224    INFO <general>: CPythonInvoker(28, C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\default.py): script successfully run
2020-10-17 16:26:02.309 T:14224    INFO <general>: Python interpreter stopped
2020-10-17 16:26:03.317 T:14132   ERROR <general>: CImageLoader::DoWork - Direct texture file loading failed for resource://resource.images.studios.white/Unspecified.png
2020-10-17 16:26:09.231 T:7932  WARNING <general>: xbmc.translatePath is deprecated and might be removed in future kodi versions. Please use xbmcvfs.translatePath instead.
2020-10-17 16:26:12.099 T:17292    INFO <general>: initializing python engine.
2020-10-17 16:26:12.310 T:17292 WARNING <general>: xbmc.translatePath is deprecated and might be removed in future kodi versions. Please use xbmcvfs.translatePath instead.
2020-10-17 16:26:12.310 T:17292 WARNING <general>: xbmc.translatePath is deprecated and might be removed in future kodi versions. Please use xbmcvfs.translatePath instead.
2020-10-17 16:26:12.597 T:1552     INFO <general>: Loading skin file: DialogConfirm.xml, load type: KEEP_IN_MEMORY
2020-10-17 16:26:15.664 T:1552     INFO <general>: Loading skin file: DialogKeyboard.xml, load type: KEEP_IN_MEMORY
2020-10-17 16:26:22.618 T:17292   ERROR <general>: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                                    - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                                   Error Type: <class 'KeyError'>
                                                   Error Contents: 'q'
                                                   Traceback (most recent call last):
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\default.py", line 5, in <module>
                                                       addon.run()
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\resources\lib\addon.py", line 17, in run
                                                       plugin.run()
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\script.module.routing\lib\routing.py", line 130, in run
                                                       self._dispatch(self.path)
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\script.module.routing\lib\routing.py", line 141, in _dispatch
                                                       view_func(**kwargs)
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\resources\lib\addon.py", line 340, in play
                                                       helper.play(plugin.args['video_id'][0])
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\resources\lib\kodihelper.py", line 212, in play
                                                       login_token = self.get_token()
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\resources\lib\kodihelper.py", line 117, in get_token
                                                       self.set_login_credentials()
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\resources\lib\kodihelper.py", line 107, in set_login_credentials
                                                       return self.set_tv_provider_credentials()
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\resources\lib\kodihelper.py", line 138, in set_tv_provider_credentials
                                                       username = self.get_user_input(username_type)
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\resources\lib\kodihelper.py", line 74, in get_user_input
                                                       self.log('User input string: {q}'.format(query))
                                                   KeyError: 'q'
                                                   -->End of Python script error report<--
                                                   
2020-10-17 16:26:22.898 T:17292    INFO <general>: Python interpreter stopped
2020-10-17 16:26:22.929 T:1552    ERROR <general>: Playlist Player: skipping unplayable item: 0, path [plugin://plugin.video.cmore/play?video_id=13290214]
2020-10-17 16:26:23.346 T:1552     INFO <general>: Loading skin file: DialogTextViewer.xml, load type: KEEP_IN_MEMORY
2020-10-17 16:26:24.258 T:7932  WARNING <general>: xbmc.translatePath is deprecated and might be removed in future kodi versions. Please use xbmcvfs.translatePath instead.
2020-10-17 16:26:32.916 T:14452    INFO <general>: initializing python engine.
2020-10-17 16:26:33.134 T:14452 WARNING <general>: xbmc.translatePath is deprecated and might be removed in future kodi versions. Please use xbmcvfs.translatePath instead.
2020-10-17 16:26:33.134 T:14452 WARNING <general>: xbmc.translatePath is deprecated and might be removed in future kodi versions. Please use xbmcvfs.translatePath instead.
2020-10-17 16:26:33.248 T:14452    INFO <general>: CPythonInvoker(30, C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\default.py): script successfully run
2020-10-17 16:26:33.506 T:14452    INFO <general>: Python interpreter stopped

kjerstadius committed:
Add missing variable reference

Doh!

kjerstadius committed:
Change Kodi Python API version

kjerstadius commented:
That issue should be fixed now and you shouldn't need to modify addon.xml either.

rikardfries commented:
> That issue should be fixed now and you shouldn't need to modify addon.xml either.

Python version 3.5.0 does not work. 3.0.0 works.
I do not know if that is the correct value though.

When I have inserted the kundnummer I get this. 
Thanks for helping me!


2020-10-21 19:43:11.676 T:14164   ERROR <general>: Control 55 in window 10025 has been asked to focus, but it can't
2020-10-21 19:43:11.691 T:9528     INFO <general>: initializing python engine.
2020-10-21 19:43:11.889 T:9528  WARNING <general>: xbmc.translatePath is deprecated and might be removed in future kodi versions. Please use xbmcvfs.translatePath instead.
2020-10-21 19:43:11.889 T:9528  WARNING <general>: xbmc.translatePath is deprecated and might be removed in future kodi versions. Please use xbmcvfs.translatePath instead.
2020-10-21 19:43:11.901 T:9528     INFO <general>: CPythonInvoker(25, C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\default.py): script successfully run
2020-10-21 19:43:12.186 T:9528     INFO <general>: Python interpreter stopped
2020-10-21 19:43:17.906 T:4656     INFO <general>: initializing python engine.
2020-10-21 19:43:18.126 T:4656  WARNING <general>: xbmc.translatePath is deprecated and might be removed in future kodi versions. Please use xbmcvfs.translatePath instead.
2020-10-21 19:43:18.127 T:4656  WARNING <general>: xbmc.translatePath is deprecated and might be removed in future kodi versions. Please use xbmcvfs.translatePath instead.
2020-10-21 19:43:18.795 T:4656     INFO <general>: CPythonInvoker(26, C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\default.py): script successfully run
2020-10-21 19:43:19.070 T:4656     INFO <general>: Python interpreter stopped
2020-10-21 19:43:20.076 T:4852  WARNING <general>: xbmc.translatePath is deprecated and might be removed in future kodi versions. Please use xbmcvfs.translatePath instead.
2020-10-21 19:43:20.871 T:6864     INFO <general>: initializing python engine.
2020-10-21 19:43:21.046 T:6864  WARNING <general>: xbmc.translatePath is deprecated and might be removed in future kodi versions. Please use xbmcvfs.translatePath instead.
2020-10-21 19:43:21.046 T:6864  WARNING <general>: xbmc.translatePath is deprecated and might be removed in future kodi versions. Please use xbmcvfs.translatePath instead.
2020-10-21 19:43:21.351 T:14164    INFO <general>: Loading skin file: DialogConfirm.xml, load type: KEEP_IN_MEMORY
2020-10-21 19:43:24.102 T:14164    INFO <general>: Loading skin file: DialogKeyboard.xml, load type: KEEP_IN_MEMORY
2020-10-21 19:43:35.101 T:4852  WARNING <general>: xbmc.translatePath is deprecated and might be removed in future kodi versions. Please use xbmcvfs.translatePath instead.
2020-10-21 19:43:50.122 T:4852  WARNING <general>: xbmc.translatePath is deprecated and might be removed in future kodi versions. Please use xbmcvfs.translatePath instead.
2020-10-21 19:44:05.147 T:4852  WARNING <general>: xbmc.translatePath is deprecated and might be removed in future kodi versions. Please use xbmcvfs.translatePath instead.
2020-10-21 19:44:12.055 T:6864    ERROR <general>: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                                    - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                                   Error Type: <class 'TypeError'>
                                                   Error Contents: argument 1 must be str, not bytes
                                                   Traceback (most recent call last):
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\default.py", line 5, in <module>
                                                       addon.run()
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\resources\lib\addon.py", line 17, in run
                                                       plugin.run()
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\script.module.routing\lib\routing.py", line 130, in run
                                                       self._dispatch(self.path)
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\script.module.routing\lib\routing.py", line 141, in _dispatch
                                                       view_func(**kwargs)
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\resources\lib\addon.py", line 340, in play
                                                       helper.play(plugin.args['video_id'][0])
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\resources\lib\kodihelper.py", line 212, in play
                                                       login_token = self.get_token()
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\resources\lib\kodihelper.py", line 117, in get_token
                                                       self.set_login_credentials()
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\resources\lib\kodihelper.py", line 107, in set_login_credentials
                                                       return self.set_tv_provider_credentials()
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\resources\lib\kodihelper.py", line 138, in set_tv_provider_credentials
                                                       username = self.get_user_input(username_type)
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\resources\lib\kodihelper.py", line 74, in get_user_input
                                                       self.log('User input string: {q}'.format(q=query))
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\resources\lib\kodihelper.py", line 53, in log
                                                       xbmc.log(msg=msg.encode('utf-8'), level=xbmc.LOGDEBUG)
                                                   TypeError: argument 1 must be str, not bytes
                                                   -->End of Python script error report<--
                                                   
2020-10-21 19:44:12.363 T:6864     INFO <general>: Python interpreter stopped
2020-10-21 19:44:12.368 T:14164    INFO <general>: Loading skin file: DialogTextViewer.xml, load type: KEEP_IN_MEMORY
2020-10-21 19:44:20.100 T:14164   ERROR <general>: Playlist Player: skipping unplayable item: 0, path [plugin://plugin.video.cmore/play?video_id=3351495]
2020-10-21 19:44:20.285 T:4852  WARNING <general>: xbmc.translatePath is deprecated and might be removed in future kodi versions. Please use xbmcvfs.translatePath instead.

kjerstadius commented:
I'm sorry for the delay @rikardfries, I'm afraid I haven't had any time to devote to this lately. The good news is that I'm working on setting up a Kodi 19 development system so I can test and verify my work locally. Hopefully it shouldn't be too long, but I don't want to make any promises.

Cheers

rikardfries commented:
> I'm sorry for the delay @rikardfries, I'm afraid I haven't had any time to devote to this lately. The good news is that I'm working on setting up a Kodi 19 development system so I can test and verify my work locally. Hopefully it shouldn't be too long, but I don't want to make any promises.
> 
> Cheers

No worries, I am thankful that you try to support me/us. 
Just go as best as you can but do not get lost trying to catch time. 
It never works 😀

Leatherface75 commented:
> I'm sorry for the delay @rikardfries, I'm afraid I haven't had any time to devote to this lately. The good news is that I'm working on setting up a Kodi 19 development system so I can test and verify my work locally. Hopefully it shouldn't be too long, but I don't want to make any promises.
> 
> Cheers

When you have that fixed maybe you can take a look on viaplay addon that also needs to be fixed?

kjerstadius committed:
Update inputstream property name

The 'inputstreamaddon' name is deprecated as of Kodi 19.

kjerstadius committed:
Remove unnecessary encode() calls

In Python 2 (I assume) these calls were used in order to ensure that the
strings were unicode, textual representations. However, in Python 3, the
encode() call converts the string to a bytes type, which goes against
the original intention and instead breaks the usage of the strings.

kjerstadius commented:
Hi @rikardfries. I finally found some time to dedicate to this and with my latest changes the plugin seems to be working in a Python 3 environment. At least it works on my Kodi 19 test device (RPi2 running LibreELEC nightly builds). I force pushed this update, so you might find you'll need to hard reset your local python3 branch by doing `git reset --hard origin/python3`.

I really hope this works for you, but please keep me updated either way.

kjerstadius commented:
> When you have that fixed maybe you can take a look on viaplay addon that also needs to be fixed?

I'm sorry to say that at present and for the foreseeable future that's unlikely as I have very limited time and no Viaplay subscription.

rikardfries commented:
> Hi @rikardfries. I finally found some time to dedicate to this and with my latest changes the plugin seems to be working in a Python 3 environment. At least it works on my Kodi 19 test device (RPi2 running LibreELEC nightly builds). I force pushed this update, so you might find you'll need to hard reset your local python3 branch by doing `git reset --hard origin/python3`.
> 
> I really hope this works for you, but please keep me updated either way.

I am sorry to say that it is not working.
I have attached my kodi log with trace on.

Looks like I have some connection problem.
When I have put in my Comhem data it crashes.
There is no hurry here on my side.
When you have time, take a look at it.
I am on windows 10 2004 with  todays kodi 19.
[2021_0109_kodi (2).zip](https://github.com/emilsvennesson/kodi-cmore/files/5792389/2021_0109_kodi.2.zip)

kjerstadius commented:
Hmm. The plugin can't resolve the operator login URL for some reason. I'm not sure whether this has anything to do with my updates, or whether CMore has changed to operator login procedure (which I personally don't use and therefore hadn't tested). While I don't get the exact same error as you, when using some dummy credentials I get one that is similar enough to try and debug from. I'll see what I can come up with.

Also, I suspect that the log contains your login credentials, so please have a look and blank them out if that's the case (I'm pretty sure they're there).

rikardfries commented:
> Hmm. The plugin can't resolve the operator login URL for some reason. I'm not sure whether this has anything to do with my updates, or whether CMore has changed to operator login procedure (which I personally don't use and therefore hadn't tested). While I don't get the exact same error as you, when using some dummy credentials I get one that is similar enough to try and debug from. I'll see what I can come up with.
> 
> Also, I suspect that the log contains your login credentials, so please have a look and blank them out if that's the case (I'm pretty sure they're there).

I have removed it for now.
Will put it back in a while.

kjerstadius commented:
I'm wondering whether CMore has simply disabled direct operator login, forcing you to connect your operator credentials to a regular CMore account. It's not possible to login with operator credentials in the Android app anymore, it will just direct you to [this page](https://www.cmore.se/login/operator), where you associate your operator credentials with a CMore account. CMore revamped their login procedure back in 2019. Back then one of the actual developers at TV4/CMore chimed in regarding the changes (in [issue 33](https://github.com/emilsvennesson/kodi-cmore/issues/33)) and said that

```
As for operator login, the typical flow would be that the user activates their account once to be able to login using regular C More credentials.
```

It seems the old style, or "direct", operator login continued to work for some time, but since the URL used for that purpose no longer seems to exist, I'm inclined to believe that at some point they simply disabled that type of login. Since it's not possible to login using operator credentials in the app anymore I'm also not able to observe a working case as I would normally do, so I don't really know how to proceed, other than to recommend the procedure mentioned earlier.

rikardfries commented:
> I'm wondering whether CMore has simply disabled direct operator login, forcing you to connect your operator credentials to a regular CMore account. It's not possible to login with operator credentials in the Android app anymore, it will just direct you to [this page](https://www.cmore.se/login/operator), where you associate your operator credentials with a CMore account. CMore revamped their login procedure back in 2019. Back then one of the actual developers at TV4/CMore chimed in regarding the changes (in [issue 33](https://github.com/emilsvennesson/kodi-cmore/issues/33)) and said that
> 
> ```
> As for operator login, the typical flow would be that the user activates their account once to be able to login using regular C More credentials.
> ```
> 
> It seems the old style, or "direct", operator login continued to work for some time, but since the URL used for that purpose no longer seems to exist, I'm inclined to believe that at some point they simply disabled that type of login. Since it's not possible to login using operator credentials in the app anymore I'm also not able to observe a working case as I would normally do, so I don't really know how to proceed, other than to recommend the procedure mentioned earlier.

IT WORKS!
You are right. When I logon with my ’ordinary’ userid and password it works.
I know that it did not work earlier. Maybe the quarrel with comhem made them go back to the old logon way. Anyway, i am happy it seems to work fine. Kudos to you that fixed it !!! Maybe it could go to the kodi repo..

Leatherface75 commented:
> > When you have that fixed maybe you can take a look on viaplay addon that also needs to be fixed?
> 
> I'm sorry to say that at present and for the foreseeable future that's unlikely as I have very limited time and no Viaplay subscription.

Ok but you can get it for free for 2 weeks and just cancel it before that period ends :)
An addon for Telenor https://stream.telenor.se/ would be nice too but that needs more work because it doesn't exists yet :(

Leatherface75 commented:
Have been testing this now and seems to work fine.

kjerstadius commented:
Glad to hear it!

Leatherface75 commented:
But i have another problem but that's not related to your changes.
The plugin only shows max 100 items so it doesn't show all episodes of some series for example.
I tried to change 'page_size': '100' to 'page_size': '500' in cmore.py file but that didn't help.

kjerstadius commented:
> But i have another problem but that's not related to your changes.
> The plugin only shows max 100 items so it doesn't show all episodes of some series for example.
> I tried to change 'page_size': '100' to 'page_size': '500' in cmore.py file but that didn't help.

There's room for improvement in places for sure. My ambition is to tackle some of these things that I've come across myself, but I can't really give a timeline.

failure84 commented:
Thanks for the good work @kjerstadius it works pretty good!

Leatherface75 commented:
Seems to be a problem with opening live channels with plugin now.

kjerstadius commented:
> Seems to be a problem with opening live channels with plugin now.

I have confirmed that there's indeed a problem with the live channels part of the plugin. Unfortunately, I have very limited time to investigate and solve this issue for the time being. I will look into it when I can.

mostman commented:
Kodi 19.4
Error when trying to stream anything, no matter if its a movie, series, program or sports channel....

2022-06-18 14:10:01.982 T:4366    ERROR <general>: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                                    - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                                   Error Type: <class 'TypeError'>
                                                   Error Contents: byte indices must be integers or slices, not str
                                                   Traceback (most recent call last):
                                                     File "/home/my_username/.kodi/addons/plugin.video.cmore/default.py", line 5, in <module>
                                                       addon.run()
                                                     File "/home/my_username/.kodi/addons/plugin.video.cmore/resources/lib/addon.py", line 17, in run
                                                       plugin.run()
                                                     File "/home/my_username/.kodi/addons/script.module.routing/lib/routing.py", line 130, in run
                                                       self._dispatch(self.path)
                                                     File "/home/my_username/.kodi/addons/script.module.routing/lib/routing.py", line 141, in _dispatch
                                                       view_func(**kwargs)
                                                     File "/home/my_username/.kodi/addons/plugin.video.cmore/resources/lib/addon.py", line 340, in play
                                                       helper.play(plugin.args['video_id'][0])
                                                     File "/home/my_username/.kodi/addons/plugin.video.cmore/resources/lib/kodihelper.py", line 214, in play
                                                       stream = self.c.get_stream(video_id, login_token=login_token)
                                                     File "/home/my_username/.kodi/addons/plugin.video.cmore/resources/lib/cmore.py", line 168, in get_stream
                                                       url = '{playback_api}{media_uri}'.format(playback_api=init_data['envPlaybackApi'], media_uri=asset['mediaUri'])
                                                   TypeError: byte indices must be integers or slices, not str
                                                   -->End of Python script error report<--

rikardfries commented:
The same for me. Will check the log when I have time.

rikardfries commented:
Here is what I get when trying to play something..

2022-06-19 22:35:11.859 T:8260    ERROR <general>: Skipped 1 duplicate messages..
2022-06-19 22:35:11.859 T:8260    ERROR <general>: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                                    - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                                   Error Type: <class 'TypeError'>
                                                   Error Contents: byte indices must be integers or slices, not str
                                                   Traceback (most recent call last):
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\default.py", line 5, in <module>
                                                       addon.run()
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\resources\lib\addon.py", line 17, in run
                                                       plugin.run()
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\script.module.routing\lib\routing.py", line 130, in run
                                                       self._dispatch(self.path)
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\script.module.routing\lib\routing.py", line 141, in _dispatch
                                                       view_func(**kwargs)
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\resources\lib\addon.py", line 340, in play
                                                       helper.play(plugin.args['video_id'][0])
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\resources\lib\kodihelper.py", line 214, in play
                                                       stream = self.c.get_stream(video_id, login_token=login_token)
                                                     File "C:\Users\rikar\AppData\Roaming\Kodi\addons\plugin.video.cmore\resources\lib\cmore.py", line 168, in get_stream
                                                       url = '{playback_api}{media_uri}'.format(playback_api=init_data['envPlaybackApi'], media_uri=asset['mediaUri'])
                                                   TypeError: byte indices must be integers or slices, not str
                                                   -->End of Python script error report<--
                                                   
2022-06-19 22:35:11.873 T:10416    INFO <general>: CActiveAESink::OpenSink - initialize sink
2022-06-19 22:35:12.361 T:8260     INFO <general>: Python interpreter stopped
2022-06-19 22:35:12.372 T:2228    ERROR <general>: Playlist Player: skipping unplayable item: 0, path [plugin://plugin.video.cmore/play?video_id=13737275]
2022-06-19 22:35:12.572 T:2228     INFO <general>: Loading skin file: DialogTextViewer.xml, load type: KEEP_IN_MEMORY

kjerstadius commented:
It seems the playback API has changed. I'm trying to wrap my head around the new API, but I can't give any sort of ETA of when I might be able to get things working again.

rikardfries commented:
I am just happy that you are helping us. If you need help with testing please just ping me!

kjerstadius commented:
Good news! I have made quite some progress with the new API and have a somewhat hacky, but mostly working version. However, it's not based on this plugin, but rather the [Telia Play plugin](https://github.com/Ikosse/plugin.video.teliaplay-se). The reason for this is that, likely due to Telia's acquisition of C More, the new C More API is mostly the same as the API used for Telia Play, albeit with C More specific URIs, some additional quirks and a different login process.

Anyway, enough technical details... If you want to give the new plugin a try, the easiest way is to do the following:
* Install the Telia Play plugin from the official Kodi addon repository
* Clone the [`feat/cmore branch`](https://github.com/kjerstadius/plugin.video.teliaplay-se/tree/feat/cmore) of [my fork](https://github.com/kjerstadius/plugin.video.teliaplay-se/tree/feat/cmore) of the Telia Play plugin
* Copy the files in `resources/lib` from my fork to `.kodi/addons/plugin.video.teliaplay-se/resources/lib`, thereby replacing the original files
* Start plugin
* Enter login info
* Profit

As I said, this is a bit of hack and there are bugs. On the other hand, most things seem to work. I have tested starting movies, series episodes, sports events, TV4 programs as well as live TV and that all works. Searching also works. Here are a couple of known issues:
* Some menu pages don't open correctly and lead to an error. Not sure why yet.
* Due to the way the original plugin is written, the country is hardcoded to `se` (Sweden) in multiple places. This likely means that it doesn't work for non-Swedish users, though I am unable to say for sure. Still, here's no harm in trying.

Please give it a try and let me know how it works out.

rikardfries commented:
It sounded a bit to good to be true but in fact it works at least as good as the old plugin.
Have watched a full show and tested all live channels. TV2 seems to be missing.. ?
Two questions:
1) Do you want me to report errors in here or at another spot ?
2) Will you create a new plugin or just overwrite the old one with this and/or will you try cooperate with the creator of the Telia Play addon ?

and of course Thank you once again for excellent work.!!!!!!
/Rikard ¤ kind regards

kjerstadius commented:
Glad to hear it! I have run into another issue with regards to refreshing the authentication token and I've pushed a fix to my fork. Please update the next time you use the plugin. Should this fix not work for you for some reason, you can always go into `.kodi/userdata/addon_data/plugin.video.teliaplay-se/`, delete `userdata.json` and then run the plugin again. That should re-login from a clean state and I used this as a workaround until I identified the proper API.

As for your questions:
1. I have enabled issues on my Telia Play plugin fork; please report issues there.
2. I have opened [an issue for the Telia Play plugin](https://github.com/Ikosse/plugin.video.teliaplay-se/issues/9) regarding what the best way forward is. My current opinion is that it might be best to create a separate plugin for C More. In that case I will rename my fork, essentially creating a new plugin.

rikardfries commented:
Funny that you should said that. Me too. :)
I will try your update tonight or tomorrow, depending on time....
Think you are right on number two, SEEMORE ?
